### PR TITLE
Ensure gc.collect() is called before checking refcount in tests.

### PR DESCRIPTION
**Pull request by sklam**
Splitted from https://github.com/numba/numba/pull/8776#issuecomment-1442164785 as requested by reviewer.

### CODE REVIEW

File: numba/tests/support.py [hunk 1]
```diff
@@ -266,6 +266,16 @@ def assertRefCount(self, *objects):
...
 
+    def assertRefCountEqual(self, *objects):
+        gc.collect()
+        rc = [sys.getrefcount(x) for x in objects]
+        for i, x in enumerate(objects[1:], start=1):
+            rc_0 = rc[0]
+            rc_i = rc[i]
+            if rc_0 != rc_i:
+                self.fail(f"Refcount for objects does not match. "
+                          f"#0({rc_0}) != #{i}({rc_i}) does not match.")
```
Comment (stuartarchibald) [inline numba/tests/support.py:277]:
```suggestion
        rc_0 = rc[0]
        for i in range(1, len(objects) - 1):
            rc_i = rc[i]
            if rc_0 != rc_i:
                self.fail(f"Refcount for objects does not match. "
                          f"#0({rc_0}) != #{i}({rc_i}) does not match.")
```
hoist loop invariant and don't think `enumerate` is needed?

File: numba/tests/test_np_randomgen.py [hunk 2]
```diff
@@ -231,9 +232,11 @@ def test_npgen_boxing_refcount(self):
         do_box = numba.njit(lambda x:x)
 
         y = do_box(rng_instance)
+        gc.collect()
```
Comment (stuartarchibald) [inline numba/tests/test_np_randomgen.py:235]: Could this use `with self.assertRefCount(...):` ?

**Review by stuartarchibald — COMMENTED** [review]
Couple of minor things else looks good, thanks for moving this patch to a separate PR @sklam.

File: numba/tests/test_np_randomgen.py (hunk 2 above)
Comment (sklam) [reply numba/tests/test_np_randomgen.py:235]: it's expecting a `+ 1` so it cannot use that testmethod

Comment (stuartarchibald) [reply numba/tests/test_np_randomgen.py:235]: ah ok, thanks for checking.

File: numba/tests/support.py (hunk 1 above)
Comment (sklam) [reply numba/tests/support.py]: shouldn't be `len(objects) - 1`.  addressed in https://github.com/numba/numba/pull/8781/commits/70cedbc9de634ef8e7f50c8cd974fb3113a8e69e

File: numba/tests/support.py [hunk 3]
```diff
@@ -266,6 +266,16 @@ def assertRefCount(self, *objects):
                 self.fail("Refcount changed from %d to %d for object: %r"
                           % (old, new, obj))
 
+    def assertRefCountEqual(self, *objects):
+        gc.collect()
+        rc = [sys.getrefcount(x) for x in objects]
+        rc_0 = rc[0]
+        for i in range(len(objects))[1:]:
```
Comment (stuartarchibald) [inline numba/tests/support.py:273]:
```suggestion
        for i in range(1, len(objects)):
```
is it just this?

Comment (sklam) [reply numba/tests/support.py:273]: yup. you can imagine `range()` returning a list. that's allowed use from py2 when range() does return a list.

Comment (stuartarchibald) [reply numba/tests/support.py:273]: Can leave it, it's just a preference.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes.

**Pull request merged**